Adding BitNet b1.58 Model

shumingma opened the pull request:
# What does this PR do?

This PR adds the support of codes for the coming BitNet b1.58 model from Microsoft. @ArthurZucker 

https://arxiv.org/abs/2402.17764

shumingma committed:
Adding BitNet b1.58 Model

shumingma committed:
Add testing code for BitNet

shumingma commented:
> Hi @shumingma, very cool project 🔥 ! it seems to me like a llama architecture with bitnet layers, did you try just using or adapting the bitnet quantization [here](https://github.com/huggingface/transformers/blob/main/src/transformers/integrations/bitnet.py) to load the model ?

Hi @MekkCyber glad you like the project, and thanks for the suggestion regarding the transformers BitNet implementation!

We did look into the transformers BitNet implementation. However, our BitNet b1.58 architecture includes some modifications, notably using subln which introduces two additional LayerNorm layers. This structural difference prevents a direct drop-in replacement using the existing transformers quantization.

Another important factor for us was designing a solution that works seamlessly for both training (online quantization) and inference (offline quantization). To achieve that flexibility and accommodate the architectural tweaks, we ended up building our own layers.

shumingma committed:
Fix format issues

shumingma committed:
Fix docstring format issues

shumingma committed:
Fix docstring

shumingma committed:
Fix docstring

shumingma committed:
Fix: weight back to uint8

shumingma committed:
Fix

shumingma committed:
Fix format issues

shumingma committed:
Remove copy comments

shumingma committed:
Add model link to the docstring

shumingma committed:
Fix: set tie_word_embeddings default to false

shumingma committed:
Merge branch 'main' into bitnet

shumingma commented:
Hi @MekkCyber Just wanted to let you know that the PR has passed all tests. Would you be willing to review the code when you have a moment? And let us know if you have any further comments or suggestions for the PR. We appreciate your input!

MekkCyber commented on src/transformers/models/bitnet/modeling_bitnet.py:
```diff
@@ -0,0 +1,1182 @@
+        self.config = config
+        self.hidden_size = config.hidden_size
+        self.intermediate_size = config.intermediate_size
+        self.gate_proj = BitLinear(
+            self.hidden_size, self.intermediate_size, bias=False, online_quant=config.online_quant
+        )
+        self.up_proj = BitLinear(
+            self.hidden_size, self.intermediate_size, bias=False, online_quant=config.online_quant
+        )
+        self.down_proj = BitLinear(
+            self.intermediate_size, self.hidden_size, bias=False, online_quant=config.online_quant
```
We can use `Linear` here

MekkCyber commented on src/transformers/models/bitnet/modeling_bitnet.py:
```diff
@@ -0,0 +1,1182 @@
+    return hidden_states.reshape(batch, num_key_value_heads * n_rep, slen, head_dim)
+
+
+class BitNetAttention(nn.Module):
+    """Multi-headed attention from 'Attention Is All You Need' paper"""
+
+    def __init__(self, config: BitNetModelConfig, layer_idx: Optional[int] = None):
+        super().__init__()
+        self.config = config
+        self.layer_idx = layer_idx
+        if layer_idx is None:
+            logger.warning_once(
+                f"Instantiating {self.__class__.__name__} without passing a `layer_idx` is not recommended and will "
+                "lead to errors during the forward call if caching is used. Please make sure to provide a `layer_idx` "
+                "when creating this class."
+            )
+
```
we moved away from this attention implementation using classes, we are adopting an attention interface now, it's far simpler. You can refer to the `llama` implementation here : https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L219C5-L290C41

MekkCyber commented on src/transformers/models/bitnet/modeling_bitnet.py:
```diff
@@ -0,0 +1,1182 @@
+        if self._use_flash_attention_2:
+            # 2d mask is passed through the layers
+            attention_mask = attention_mask if (attention_mask is not None and 0 in attention_mask) else None
+        elif self._use_sdpa and not output_attentions:
+            # output_attentions=True can not be supported when using SDPA, and we fall back on
+            # the manual implementation that requires a 4D causal mask in all cases.
+            attention_mask = _prepare_4d_causal_attention_mask_for_sdpa(
+                attention_mask,
+                (batch_size, seq_length),
+                inputs_embeds,
+                past_key_values_length,
+            )
+        else:
+            # 4d mask is passed through the layers
+            attention_mask = _prepare_4d_causal_attention_mask(
+                attention_mask, (batch_size, seq_length), inputs_embeds, past_key_values_length
```
Instead of this implementation we are using an _update_causal_mask method to handle that : https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L608

MekkCyber reviewed (commented):
Hey @shumingma, I left a few comments. Since the architecture is quite similar to LLaMA, I think it would be a great fit for the new [modular Transformers](https://huggingface.co/docs/transformers/en/modular_transformers) setup.

You could simply extend the existing classes like `LlamaRMSNorm`, `LlamaAttention`, etc., in the modular file, and use the converter to handle the modeling script. If you have any questions or need help with something please let me know

shumingma committed:
Update

Commit:
Generate modeling file

shumingma committed:
Change config name for automatically generating modeling file.

Commit:
Generate modeling file

shumingma committed:
Fix class name

shumingma committed:
Change testing branch

shumingma committed:
Remove unused param

shumingma committed:
Fix config docstring

MekkCyber commented on src/transformers/models/bitnet/modular_bitnet.py:
```diff
@@ -0,0 +1,175 @@
+_CHECKPOINT_FOR_DOC = "microsoft/bitnet-b1.58-2B-4T"
+
+
+class BitNetRMSNorm(LlamaRMSNorm):
+    pass
+
+
+class BitNetMLP(GemmaMLP):
+    def __init__(self, config: BitNetConfig):
+        super().__init__(config)
+        self.ffn_sub_norm = BitNetRMSNorm(config.intermediate_size, eps=config.rms_norm_eps)
```
Thanks for using the modular

MekkCyber commented on src/transformers/utils/quantization_config.py:
```diff
@@ -1691,14 +1691,18 @@ def from_dict(cls, config_dict, return_unused_kwargs=False, **kwargs):
 @dataclass
-class BitNetConfig(QuantizationConfigMixin):
+class BitNetQuantConfig(QuantizationConfigMixin):
     def __init__(
         self,
         modules_to_not_convert: Optional[List] = None,
+        online_quantization: bool = False,
+        offline_quantization: bool = False,
```
can you add a docstring please to explain what is online_quantization and offline_quantization

shumingma replied:
done

MekkCyber reviewed (approved):
LGTM 🤗 ! Thanks a lot for iterating ! can you make sure to add the quantization config to the models on the hub, and test if the model is generating as expected ?

shumingma committed:
Add docstring for BitNetQuantConfig.

shumingma committed:
Fix docstring

shumingma commented:
> LGTM 🤗 ! Thanks a lot for iterating ! can you make sure to add the quantization config to the models on the hub, and test if the model is generating as expected ?

Thanks! I've updated the config on the Hub: [https://huggingface.co/microsoft/bitnet-b1.58-2B-4T/blob/main/config.json#L26-L29](https://www.google.com/url?sa=E&q=https%3A%2F%2Fhuggingface.co%2Fmicrosoft%2Fbitnet-b1.58-2B-4T%2Fblob%2Fmain%2Fconfig.json%23L26-L29) and confirmed the generation behavior is as expected.

shumingma committed:
Merge branch 'main' into bitnet

MekkCyber commented on src/transformers/models/bitnet/modular_bitnet.py:
```diff
@@ -0,0 +1,175 @@
+        return attn_output, attn_weights
+
+
+class BitNetDecoderLayer(LlamaDecoderLayer):
+    def __init__(self, config: BitNetConfig, layer_idx: int):
+        super().__init__()
+        self.self_attn = BitNetAttention(config=config, layer_idx=layer_idx)
+        self.mlp = BitNetMLP(config)
```
just FMI, was adding the `self.self_attn` and `self.mlp` necessary ? they were non added automatically with the `rmsnorm` layers ?

Cyrilvallez replied:
This is not needed and will be picked-up automatically by modular -> to remove

MekkCyber reviewed (approved):
Did a quick last pass ! left some small comments for improvements, but it's good on my side ! pinging @SunMarc or @Cyrilvallez to review

MekkCyber commented on docs/source/en/model_doc/bitnet.md:
```diff
@@ -0,0 +1,124 @@
+
+Trained on a corpus of 4 trillion tokens, this model demonstrates that native 1-bit LLMs can achieve performance comparable to leading open-weight, full-precision models of similar size, while offering substantial advantages in computational efficiency (memory, energy, latency).
+
+➡️ **Technical Report:** [BitNet b1.58 2B4T Technical Report](https://arxiv.org)
```
```suggestion
➡️ **Technical Report:** [BitNet b1.58 2B4T Technical Report](https://arxiv.org/abs/2504.12285)
```

SunMarc commented on docs/source/en/model_doc/bitnet.md:
```diff
@@ -0,0 +1,124 @@
+
+**VERY IMPORTANT NOTE ON EFFICIENCY**
+
+> Please do NOT expect performance efficiency gains (in terms of speed, latency, or energy consumption) when using this model with the standard transformers library, even with the required fork.
```
is it still recommended to use the fork after this PR gets merged ? 

```suggestion
> Please do NOT expect performance efficiency gains (in terms of speed, latency, or energy consumption) when using this model with the standard transformers library.
```

SunMarc commented on docs/source/en/model_doc/bitnet.md:
````diff
@@ -0,0 +1,124 @@
+
+### Requirements
+
+```bash
+pip install git+https://github.com/shumingma/transformers.git
+```
+We are actively working with the Hugging Face team to integrate the necessary code into the main `transformers` library. This installation method may change in the future.
````
Update this

SunMarc commented on src/transformers/integrations/bitnet.py:
```diff
@@ -190,6 +190,84 @@ def forward(self, input):
+class WeightQuant(torch.autograd.Function):
+    @staticmethod
+    @torch.compile
+    def forward(ctx, x):
```
let's add a small explanation of what is happening here

SunMarc commented on src/transformers/integrations/bitnet.py:
```diff
@@ -190,6 +190,84 @@ def forward(self, input):
+class ActQuant(torch.autograd.Function):
+    @staticmethod
+    @torch.compile
+    def forward(ctx, x):
```
same here

SunMarc commented on src/transformers/integrations/bitnet.py:
```diff
@@ -190,6 +190,84 @@ def forward(self, input):
+            state_dict[prefix + "weight"] = unpack_weights(state_dict[prefix + "weight"], dtype=self.weight.dtype)
+        return state_dict
+
+    def forward(self, input):
+        if self.online_quant:
+            weight = WeightQuant.apply(self.weight)
+        else:
+            weight = self.weight
+        input = ActQuant.apply(input)
```
let's also add a quick comment to explain what happens here

SunMarc commented on src/transformers/integrations/bitnet.py:
```diff
@@ -190,6 +190,84 @@ def forward(self, input):
+            weight = WeightQuant.apply(self.weight)
+        else:
+            weight = self.weight
+        input = ActQuant.apply(input)
+        output = F.linear(input, weight, self.bias)
```
yeah there is room for improvement with dedicated kernels in the future potentially to speedup the computation of the 8bit activation and the ternary weights

SunMarc commented on src/transformers/integrations/bitnet.py:
```diff
@@ -190,6 +190,84 @@ def forward(self, input):
+        return grad_input
+
+
+class AutoBitLinear(nn.Linear):
+    def __init__(
+        self,
+        in_features: int,
+        out_features: int,
+        bias: bool = True,
+        device=None,
+        dtype=None,
+        online_quant: bool = False,
+    ):
+        super(AutoBitLinear, self).__init__(in_features, out_features, bias)
+        self.online_quant = online_quant
+        if not online_quant:
+            self.register_buffer(
+                "weight_scale",
+                torch.ones(
+                    (1),
+                    dtype=dtype,
+                    device=device,
+                ),
+            )
+            self._register_load_state_dict_pre_hook(self.load_hook)
```
I don't mind having a separate linear implementation if bitnetlinear is totally incompatible / breaking change. Could you explain a bit the reason ? Otherwise, I think that it is better upstream the changes you did here to BitNetLinear. cc @MekkCyber

shumingma replied:
The main issue is that the current BitLinear lacks online quantization (needed for training) and it unpacks weights during the forward pass. Unpacking repeatedly for every generated token severely slows down inference.

Our PR unpacks weights once at model load time, making generation significantly faster. This difference in handling weights, especially for inference efficiency, motivates the changes.

SunMarc commented on src/transformers/integrations/bitnet.py:
```diff
@@ -218,15 +296,35 @@ def _replace_with_bitnet_linear(
-                        device=module.weight.device,
-                        dtype=module.weight.dtype,
-                    )
+                    if quantization_config and quantization_config.online_quantization:
```
I think it will be better to create a new arg called `linear_class` instead of relying on `online_quantization` if we decide to keep two separate linear impl.

SunMarc commented on src/transformers/integrations/bitnet.py:
```diff
@@ -218,15 +296,35 @@ def _replace_with_bitnet_linear(
+                            dtype=module.weight.dtype,
+                            online_quant=True,
+                        )
+                    elif quantization_config and quantization_config.offline_quantization:
```
instead of having `online_quantization` and `offline_quantization`, let's do a single attribute, `quantization_config.use_fake_quant = True` or `quantization_config.enable_qat` = True`. Feels like it is easier to understand. If you prefer `online` of `offline` term, we can also do `quantization_mode` = "online"/ "offline"

SunMarc commented on src/transformers/integrations/bitnet.py:
```diff
@@ -190,6 +190,84 @@ def forward(self, input):
+            )
+            self._register_load_state_dict_pre_hook(self.load_hook)
+
+    def load_hook(
+        self,
+        state_dict,
+        prefix,
+        *args,
+        **kwargs,
+    ):
+        if (prefix + "weight") in state_dict and state_dict[prefix + "weight"].dtype != self.weight.dtype:
+            state_dict[prefix + "weight"] = unpack_weights(state_dict[prefix + "weight"], dtype=self.weight.dtype)
+        return state_dict
```
why we need to `unpack` the state dict here ?

shumingma replied:
To optimize inference speed. We unpack the weights upfront during loading to avoid the performance bottleneck of repeatedly unpacking them in the forward pass during token generation.

SunMarc reviewed (commented):
Thanks for upstreaming this to transformers, really like the work you guys do! On the quantization side, the changes looks great, I only have a few minor questions and we are good the merge. I'll let @Cyrilvallez comment on the modeling side.

shumingma committed:
Update docs/source/en/model_doc/bitnet.md

Co-authored-by: Mohamed Mekkouri <93391238+MekkCyber@users.noreply.github.com>

shumingma committed:
Update docs/source/en/model_doc/bitnet.md

Co-authored-by: Marc Sun <57196510+SunMarc@users.noreply.github.com>

shumingma committed:
Update bitnet config

shumingma commented:
Hi @MekkCyber @SunMarc 

Thank you for your feedback on the previous version. I've addressed the comments and suggestions you provided and have pushed the updated code.

Could you please review the changes again at your convenience?

Thanks again for your time and input!

SunMarc reviewed (approved):
Thanks a lot iterating ! LGTM for the quantization part ! Just a few minor nits

shumingma committed:
Update explanation between online and offline mode

shumingma committed:
Remove space

shumingma commented:
Hi @Cyrilvallez 

Could you please take a look at this PR when you have some availability? It's ready for merging pending your review. 

Let me know if anything needs clarification. 

Thanks!

SunMarc committed:
Merge branch 'main' into bitnet

Cyrilvallez commented on .circleci/create_circleci_config.py:
```diff
@@ -16,10 +16,9 @@
 import os
-import random
 from dataclasses import dataclass
 from typing import Any, Dict, List, Optional
-import glob
+
```
We should revert all changes to this file!

Cyrilvallez commented on .circleci/parse_test_outputs.py:
```diff
@@ -1,5 +1,6 @@
-import re
 import argparse
+import re
+
 
```
Same here

Cyrilvallez commented on .github/scripts/assign_reviewers.py:
```diff
@@ -13,14 +13,16 @@
 
-import os
-import github
 import json
```
And here

Cyrilvallez commented on benchmark/benchmark.py:
```diff
@@ -31,9 +31,7 @@
 from pathlib import Path
 
 from git import Repo
-
 from huggingface_hub import HfApi
-
```
and here

Cyrilvallez commented on benchmark/benchmarks_entrypoint.py:
```diff
@@ -2,11 +2,11 @@
 import logging
 import os
-from typing import Dict
 import sys
+from typing import Dict
```
and here

Cyrilvallez commented on benchmark/llama.py:
```diff
@@ -1,13 +1,14 @@
-from logging import Logger
 import os
+from logging import Logger
```
and here

Cyrilvallez commented on docs/source/en/model_doc/bitnet.md:
````diff
@@ -0,0 +1,124 @@
+>
+> For achieving the efficiency benefits demonstrated in the technical paper, you MUST use the dedicated C++ implementation: [bitnet.cpp](https://github.com/microsoft/BitNet).
+
+### Requirements
+
+```bash
+pip install git+https://github.com/shumingma/transformers.git
+```
+We are actively working with the Hugging Face team to integrate the necessary code into the main `transformers` library. This installation method may change in the future.
````
Should be removed

Cyrilvallez commented on scripts/benchmark/trainer-benchmark.py:
```diff
@@ -186,11 +186,11 @@ def get_base_command(args, output_dir):
     args.base_cmd = re.sub(r"[\\\n]+", " ", args.base_cmd)
 
     # remove --output_dir if any and set our own
-    args.base_cmd = re.sub("--output_dir\s+[^\s]+", "", args.base_cmd)
+    args.base_cmd = re.sub(r"--output_dir\s+[^\s]+", "", args.base_cmd)
```
We need to revert the changes to this file as well!

Cyrilvallez commented on src/transformers/integrations/bitnet.py:
```diff
@@ -190,6 +190,98 @@ def forward(self, input):
+        dtype=None,
+        online_quant: bool = False,
+    ):
+        super(AutoBitLinear, self).__init__(in_features, out_features, bias)
```
```suggestion
        super().__init__(in_features, out_features, bias)
```

Cyrilvallez commented on src/transformers/models/bitnet/__init__.py:
```diff
@@ -0,0 +1,27 @@
+# Copyright 2024 The BitNet Team and The HuggingFace Inc. team. All rights reserved.
```
2025!

Cyrilvallez commented on src/transformers/models/bitnet/configuration_bitnet.py:
```diff
@@ -0,0 +1,155 @@
+# coding=utf-8
+# Copyright 2025 Microsoft, EleutherAI and the HuggingFace Inc. team. All rights reserved.
+#
+# This code is based on EleutherAI's GPT-NeoX library and the GPT-NeoX
+# and OPT implementations in this library. It has been modified from its
+# original forms to accommodate minor architectural differences compared
+# to GPT-NeoX and OPT used by the Meta AI team that trained the model.
```
Probably wrong here

Cyrilvallez commented on src/transformers/models/bitnet/modular_bitnet.py:
```diff
@@ -0,0 +1,175 @@
+# coding=utf-8
+# Copyright 2025 Microsoft, EleutherAI and the HuggingFace Inc. team. All rights reserved.
+#
+# This code is based on EleutherAI's GPT-NeoX library and the GPT-NeoX
+# and OPT implementations in this library. It has been modified from its
+# original forms to accommodate minor architectural differences compared
+# to GPT-NeoX and OPT used by the Meta AI team that trained the model.
```
Similarly here?

Cyrilvallez commented on src/transformers/models/bitnet/modular_bitnet.py:
```diff
@@ -0,0 +1,175 @@
+class BitNetModel(LlamaModel):
+    pass
+
+
+class KwargsForCausalLM(FlashAttentionKwargs, LossKwargs): ...
```
Similarly this is not needed! Let's simply not use typing for the `super_kwargs` as it's simply sugar for the modular!

Cyrilvallez commented on src/transformers/utils/quantization_config.py:
```diff
@@ -1761,14 +1761,46 @@ def from_dict(cls, config_dict, return_unused_kwargs=False, **kwargs):
+        quantization_mode: Optional[str] = "offline",
         **kwargs,
     ):
+        assert linear_class in ["bitlinear", "autobitlinear"], (
+            f"linear_class must be either 'bitlinear' or 'autobitlinear', but got {linear_class}"
+        )
+        assert quantization_mode in ["online", "offline"], (
+            f"quantization_mode must be either 'online' or 'offline', but got {quantization_mode}"
+        )
```
Let's raise proper Exceptions here instead of assert!

Cyrilvallez commented on tests/models/bitnet/test_modeling_bitnet.py:
```diff
@@ -0,0 +1,251 @@
+        self.skipTest(reason="BitNet flash attention does not support right padding")
+
+
+@require_torch
+class BitNetIntegrationTest(unittest.TestCase):
+    @slow
+    def test_model_logits(self):
+        input_ids = [128000, 128000, 1502, 25, 2650, 527, 499, 30, 128009, 72803, 25, 220]
+        model = BitNetForCausalLM.from_pretrained("microsoft/bitnet-b1.58-2B-4T")
+        input_ids = torch.tensor([input_ids]).to(model.model.embed_tokens.weight.device)
+        with torch.no_grad():
+            out = model(input_ids).logits.float().cpu()
+        # Expected mean on dim = -1
+        EXPECTED_MEAN = torch.tensor(
+            [
+                [
+                    -1.8665,
```
I'm fine having them here as it's usually what we do for all models!

Cyrilvallez reviewed (commented):
Hey! Super nice, great work! 🤗 Just left some minor comments - basically we need to revert changes to all files that are not needed, and we can simplify a tiny bit further the modular!

MekkCyber commented:
Hi @shumingma, the PR is nearly complete — it’d be great to get it merged by the end of the week!
Could you take a look at the remaining comments, they should not be complicated?
If you’re short on time, we’re happy to help out 🤗 !

MekkCyber commented:
Thanks for the quick response, @shumingma! I’ve opened a new PR with your commits: https://github.com/huggingface/transformers/pull/37742.
I believe it’s in good shape, but feel free to review it whenever you have a moment.
We’re aiming to merge it by tomorrow at the latest.

MekkCyber commented:
Closing now that https://github.com/huggingface/transformers/pull/37742 is merged
